### DISCUSSION TITLE
Temperature correction for ultrasonic oil level gauge

### DISCUSSION THREAD

**Discussion by rowanbradley** [merbanan/rtl_433#2852] [General]
I have an Apollo (rocket shaped) oil level detector on my oil tank. I am reading its results in cm using rtl_433. But assuming that this is actually a measurement of the time taken for the ultrasonic signal to get from the sensor to the oil surface and back again, presumably I need to correct this for the changes in sound velocity with temperature, and the changes in oil density with temperature, and (maybe) the expansion of the (plastic) tank, if I want to accurately read the amount of oil (in kg) in the tank. Does anyone have any information on how to do this? Presumably the same will apply to the Watchman and other ultrasonic sensors.

Thank you - Rowan

**Comment by 50percentDave** [comment]
I know this is an old post, but I've just found the Discussions section!

I have an oil watchman and I think it does allow for the impact of temperature on the time of flight of the sound pulse (not sure), though I don't know whether it allows for oil density changes or other variables.

To accurately convert the distance in cm to the amount of oil in my tank I recorded the amount of oil to fill the tank and the cm decoded from the watchman signals over a number of tank top ups - this needs to be over a range of fill up amounts, ie from your suppliers minimum delivery, a fill up in the typical refill range, and some in the tank low range.   I then took that data and plotted a graph on https://mycurvefit.com/ to generate a best fit formula, and I use that to estimate how much oil it will take to fill my tank - after a number of readings this has proven to be be very accurate, and it gets better the more refill data you have (as accurate as it can be as the top off volume depends a bit on the refueling operator).  For me it's not unusual to predict the amount of oil required to fill up a 2500 litre tank to 50 - 75 or so litres.

If you combine the fill amount for your tank, with a daily check on oil prices (and graph them), you can determine whether the price is favourable for a top up or not.

HTH
